lib/common/http2client.c

```diff
@@ -1099,16 +1107,25 @@ static void on_connection_ready(struct st_h2o_http2client_stream_t *stream, stru
         stream->input.message_body_forbidden = 1;
     }
 
-    /* send headers */
+    /* headers */
     h2o_hpack_flatten_request(&conn->output.buf, &conn->output.header_table, conn->peer_settings.header_table_size,
                               stream->stream_id, conn->peer_settings.max_frame_size, method, &url, headers, num_headers,
                               stream->state.req == STREAM_STATE_CLOSED);
 
+    /* body */
     if (stream->state.req == STREAM_STATE_BODY) {
         h2o_buffer_init(&stream->output.buf, &h2o_socket_buffer_prototype);
         h2o_buffer_append(&stream->output.buf, body.base, body.len);
     }
     h2o_linklist_insert(&conn->output.sending_streams, &stream->output.sending_link);
+
+    /* trailers */
+    if (trailers != NULL) {
```

Comment:
```suggestion
    if (trailers != NULL && num_trailers != 0) {
        assert(stream->output.proceed_req == NULL);
```

In `do_write_req`, we make sure that `stream->output.trailers.entries` becomes non-NULL only when the list is going to be non-empty. I do not how we use `trailers`, but it should be consistent at least. Addition of `num_trailers != 0` as well as the change to the doc-comment builds on top of the requirements that we have in `do_write_req`.

Separately, am I correct to understand that the user is forbidden from providing trailers if it has opted in to use request streaming? If so, it might make sense to have an assert.

Reply:
That makes sense, addressed in https://github.com/h2o/h2o/pull/3241/commits/c61257b88d8ae3dda925b5df1854dc8eb72427e7

> Separately, am I correct to understand that the user is forbidden from providing trailers if it has opted in to use request streaming? If so, it might make sense to have an assert.

Thank you, this is a great point. Ideally we might be able to merge the passed trailers with coming request trailers, but this assertion should be sufficient for now.